### ISSUE TITLE
[Bug]: Offset only uses one window

### ISSUE THREAD

**Issue by FalcoGer** [akinsho/bufferline.nvim#789]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

When multiple offsets are defined, then opening a new window will only apply the offset configuration to one window, usually the leftmost one or the last one opened. It behaves oddly.

For example when having an offset for filetype "help" and an offset for filetype "NvimTree", then opening NvimTree will produce the correct behavior. If you then however open the help with `:vertical help`, the help window gets the offset configuration applied but the nvim-tree gets it's text deleted.

To reproduce:

1. Open nvim and only have one buffer open that doesn't have an offset setting for it's filetype
2. Open nvim-tree
![image](https://github.com/akinsho/bufferline.nvim/assets/16250196/3d5aa79b-5ecf-4108-b8ae-e8e86d10d006)

3. Open vertical help
![image](https://github.com/akinsho/bufferline.nvim/assets/16250196/fd5c8661-6d85-4aac-a542-8f001b3c013e)

4. Notice help doesn't get the title and instead the bufferline starts over the help window
5. Close nvim-tree
![image](https://github.com/akinsho/bufferline.nvim/assets/16250196/41801733-1b1c-4083-a1a6-20b48ecfd089)

6. Notice that help now gets the offset applied where before it did not

Windows that open on the right, such as tagbar, also behave oddly when opening and closing them mixed with opening and closing multiple windows that open on the left, such as undotree.

### What did you expect to happen?

Vertically split windows always get their header text and all the offsets are added for the bufferline, except of course for the offsets for windows on the right, which should be handled separately

### Config

```lua
-- ...
offsets = {
    {
        filetype = "NvimTree",
        text = function()
            return "NVIMTREE TEXT"
            -- return vim.fn.getcwd()
        end,
        -- highlight = "Directory",
        text_align = "left",
        separator = true
    },
    {
        filetype = "NERDTree",
        text = function()
            return vim.fn.getcwd()
        end,
        highlight = "Directory",
        text_align = "left",
        separator = true
    },
    {
        filetype = "undotree",
        text = "Undo Tree",
        text_align = "left",
        separator = true
    },
    {
        filetype = "diff",
        text = "Diff",
        text_align = "left",
        separator = true
    },
    {
        filetype = "tagbar",
        text = "Tags",
        text_align = "left",
        separator = true
    },
    {
        filetype = "help",
        text = "Help",
        text_align = "left",
        separator = true
    }
},
-- ...
```

### Additional Information

N/A

**Comment by akinsho** [comment]
Hi @FalcoGer,

So this is an issue I've discussed many times over the course of the history of this feature leading all the way back to when I initially added it. It is absolutely not a full proof feature that covers all use cases perfectly. The main object was for it to be good enough for 90% of people's use cases e.g. having a file explorer open etc.

The main difficulty with getting this perfectly right is that it works based on getting the window layout for nvim which comes in the form of nested array which describe the layout of neovim's windows. These array can be in a *a lot* of different configurations some of which apply to an offset and many more which don't. The offset logic just tries to find the ones that do apply and provide offsets for those.

This may or may not be easily solvable by someone taking a look and having a go. 
I'll explain my position on this. This feature to my mind is a good is better than perfect sort of thing IMO. It won't hit all cases and it's a pain to try to make it do that since as I mentioned I personally believe 90% of the use case is a single file explorer to the left.

When I implemented the feature I explained to the user who requested it at the time that it was going to be a good enough type solution not a perfect solution. I've also described this in many issues about this involving undotree et al.

I'm a very big believer in community involvement in open source projects, I think if this bothers you you are very welcome to help contribute a fix and I can help guide that. This bug is not ideal but frankly the time it might take me to dig into this or how to fit this into the many many other issues and bugs across multiple plugins I have to deal with means this is not something I'm likely to personally prioritise as it doesn't affect me in my daily work *at all*.

TLDR: PR welcome please 🙏🏾 otherwise I will leave this open for a week or two (or till I remember to check back) otherwise I will close this out.

**Comment by FalcoGer** [comment]
The problem is that I'm not very good at lua or vimscript.

However it seems the problem appears to be solvable by examining `wininfo()`, which has the fields `wincol` and `winrow`. Since you only care about `winrow == 1` you can check all windows with `winrow == 1` from `wincol == 1` up to the maximum and apply the offset thing where it matches, and then place the bufferlist in the first available gap.

**Comment by akinsho** [comment]
> The problem is that I'm not very good at lua or vimscript

@FalcoGer no vimscript knowledge is really required at all since there is not really any vimscript in this plugin and tbh lua is a very simple language. I think if you are able to configure your editor with it/know literally any programming languages that makes lua kind of trivial for this level of change.

Thanks for digging into potential solutions. Seems this mechanism could maybe simplify things massively by just doing something like
```lua
local offsets = {}
for _, win in ipairs(vim.fn.getwininfo()) do
	if win.winrow == 2 and (wincol == 1 or wincol == max) then
		table.insert(offsets, win.id)
	end
end
```

or something like this.

You could just remove this logic and the usage
https://github.com/akinsho/bufferline.nvim/blob/cd27a52ecdfed7f14a41b61b7976f155e3d593c7/lua/bufferline/offset.lua#L108-L121

`vim.fn.layout` and just add a new function like `is_offset_win(offset)`
and do a check similar to the above to return `true` or `false`

I think if you give it a try it'll be simpler than you thing probably and you can post questions here.

**Comment by FalcoGer** [comment]
It might not actually be winrow 2. probably best to get the minimum first. For example vim-airline has an option to be displayed at the top. not sure if it then goes above or below bufferline. Either way it will probably push the windows down into row 3.

**Comment by toh995** [comment]
Hi @akinsho thanks for your work on this. I noticed the PR is still unmerged - is there anything blocking the merge? Maybe I can help?

**Comment by akinsho** [comment]
There are still bugs with that issue and I don't have time to work on it now ATM
